diffrax/_solver/srk.py

```diff
@@ -0,0 +1,740 @@
+    SRK for solving additive noise SDEs.
+    """
+
+    # assuming SDE has additive noise we only need a 1-dimensional array
+    # of length s for the coefficients in front of the Brownian increment
+    # and/or Lévy areas (where s is the number of stages of the solver).
+    # This is the equivalent of the matrix a for the Brownian motion and
+    # its Lévy areas.
+    a: Float[np.ndarray, " s"]
+    b: FloatScalarLike
+
+    def check(self):
+        assert self.a.ndim == 1
+        return self.a.shape[0]
+
+
+class _AbstractGeneralCoeffs(AbstractStochasticCoeffs):
+    a: eqx.AbstractVar[tuple[np.ndarray, ...]]
+    b: eqx.AbstractVar[Float[np.ndarray, " s"]]
+
+
+class GeneralCoeffs(_AbstractGeneralCoeffs):
+    """General coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving SDEs with any type of noise (i.e. non-additive).
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        return self.b.shape[0]
+
+
+class GeneralCoeffsWithError(_AbstractGeneralCoeffs):
+    """General coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving SDEs with any type of noise (i.e. non-additive). Use this subclass
+    when the SRK provides an embedded method for error estimation.
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+    b_error: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        assert self.b_error.ndim == 1
+        assert self.b_error.shape == self.b.shape
+        return self.b.shape[0]
+
+
+COEFFS = TypeVar("COEFFS", bound=AbstractStochasticCoeffs)
+
+
+class AbstractStochasticTableau(eqx.Module, Generic[COEFFS]):
+    r"""Part of a `StochasticButcherTableau` that represents the coefficients
+    for the Brownian increment and possibly its Levy areas."""
+
+    coeffs_w: eqx.AbstractVar[COEFFS]
+
+    @property
+    def a_w(self):
+        """The coefficients in front of the Brownian increment at each stage."""
+        return self.coeffs_w.a
+
+    @property
+    def b_w(self):
+        """The coefficient for the Brownian increment when computing the output."""
+        return self.coeffs_w.b
+
+    @property
+    def get_b_error_w(self) -> Optional[Float[np.ndarray, " s"]]:
+        if isinstance(self.coeffs_w, GeneralCoeffsWithError):
+            return self.coeffs_w.b_error
+        else:
+            return None
+
+    @property
+    def additive_noise(self):
+        return isinstance(self.coeffs_w, AdditiveCoeffs)
+
+    @property
+    def has_error_estimate(self):
+        return isinstance(self.coeffs_w, GeneralCoeffsWithError)
+
+    def check(self):
```

Comment:
Nit: should be an `abc.abstractmethod`,  again a quick docstring and return annotation would be nice.